gamemodes/terrortown/entities/entities/ttt_logic_role.lua

```diff
@@ -24,35 +26,46 @@ function ENT:KeyValue(key, value)
 			value = _G[value] or value
 		end
 
-		self.Role = tonumber(value)
+		self.checkingRole = tonumber(value)
 
-		if not self.Role then
+		if not self.checkingRole then
 			ErrorNoHalt("ttt_logic_role: bad value for Role key, not a number\n")
 
-			self.Role = ROLE_NONE
+			self.checkingRole = ROLE_NONE
 		end
 	end
 end
 
 ---
--- @param string name
--- @param Entity|Player activator
--- @return[default=true] boolean
--- @realm shared
-function ENT:AcceptInput(name, activator)
+-- Called when another entity fires an event to this entity.
+-- @param string name The name of the input that was triggered
+-- @param Entity|Player activator The initial cause for the input getting triggered (e.g. the player who pushed a button)
+-- @param Entity caller The entity that directly triggered the input (e.g. the button that was pushed)
+-- @param string data The data passed
+-- @return[default=true] boolean Return true if the default action should be supressed
+-- @realm server
+function ENT:AcceptInput(name, activator, caller, data)
 	if name == "TestActivator" then
-		if IsValid(activator) and activator:IsPlayer() then
-			local activator_role = (GetRoundState() == ROUND_PREP) and ROLE_INNOCENT or activator:GetBaseRole()
+		if not IsValid(activator) or not activator:IsPlayer() then return end
+
+		---
+		-- @realm server
+		local role, team = hook.Run("TTT2ModifyLogicRoleCheck", activator, self, activator, caller, data)
+		local activatorRole = roles.GetByIndex(role):GetBaseRole()
+		local activatorTeam = (GetRoundState() == ROUND_PREP) and TEAM_INNOCENT or team
 
-			if self.Role == ROLE_NONE or self.Role == activator_role then
-				Dev(2, activator, "passed logic_role test of", self:GetName())
+		if self.checkingRole == ROLE_TRAITOR and util.IsEvilTeam(activatorTeam)
+			or self.checkingRole == ROLE_INNOCENT and not util.IsEvilTeam(activatorTeam)
+			or self.checkingRole == activatorRole and not (self.checkingRole == ROLE_TRAITOR or self.checkingRole == ROLE_INNOCENT)
```

Comment:
Why do you also check traitor and innocent again? Wouldnt those only introduce edge cases and arent they wrong anyways?

Reply:
Because otherwise this does fun things if you return `ROLE_INNOCENT, TEAM_TRAITOR`. While this is a stupid combination, it should be understandable.

Reply:
Actually i dont think this is wrong since it is only a role check and if i have the innocent role i should be included even if you return a stupid team

Reply:
One case thats a bit off, but is future-proof, would be a traitor, thats in another team, like the pirate.
He should be working for the right team and therefore count as them in a test. But because of his role he still would have access to the other stuff or count as traitor/valid even if hes in the right team.